### DISCUSSION TITLE
Stein vs. discrete-time Lyapunov equation

### DISCUSSION THREAD

**Discussion by artpelling** [pymor/pymor#1522] [Q&A] [answered]
Hello everyone,
I am looking into adding scipy and slycot bindings for discrete-time Lyapunov equations so that Gramians can be computed for discrete-time systems.
I wonder what would be the best way of naming this.

Pro "discrete Lyapunov":
- less code overall / no redundancies
- same naming convention as scipy
- as far as i can tell, only the "special case" Stein equation, where B=A^H and Q hermitian will be needed at this point.

Pro "Stein":
- @pmli suggested this and he knows much more than me :)
- if I have seen this correctly, pyMESS does not support discrete-time Lyapunov equations. At the moment, pyMESS is prioritized as default solver backend, which means that a fallback has to be implemented for the discrete case. With a Stein naming, there would be another set of solver backend defaults.

What do you think?

Another question:
In `slycot.py`, the Cholesky factors are computed with `_chol` in `solve_lyap_lrcf`. Is this better than using the Slycot routines `sb03od` and `sg03bd`?

I have created a draft PR #1525 with a "Lyapunov" naming. Its weak point being the functioning but ugly backend fallback.

**Comment by lbalicki** [comment, marked as answer]
Hi @artpelling, generally I would be for referring to an equation of the type `A X A^H - X = -Q` as the discrete-time Lyapunov equation. This would be consistent with other libraries (e.g. scipy, slicot). 

Regardless of the naming, in my opinion it would be a good idea to have separate functions for discrete- and continuous-time Lyapunov equations. I see that having just the `solve_lyap_lrcf` function with the `cont_time` argument leads to less code then having e.g. `solve_cont_lyap_lrcf` and `solve_disc_lyap_lrcf`. However, in the latter case documentation is shorter and implementation of default backends is nicer. Maybe @pmli has another opinion regarding this.

Currently [SG03AD](http://slicot.org/objects/software/shared/doc/SG03AD.html) is used for Lyapunov equations in the slycot bindings, which computes some condition estimates that `sg03bd` does not compute. If `sg03bd`,`sb03od` are significantly faster it might be worthwhile to switch to them.

**Reply by artpelling** [reply]
Thanks! I will see if I manage to implement some meaningful performance tests :)

**Reply by drittelhacker** [reply]
@lbalicki if one would want to go down the one function for all lane, I would even suggest to use`sylv` rather than `lyap` in the name and prepare it for the most general Sylvester-type equation `A X B + C X D = E` of which both equations are special cases and other cases have applications in transfer function evaluation, rational Krylov subspaces, eigenvalue problems etc. 

Regarding the SG03AD vs BD discussion: The first implements Bartels-Steward while the second is essentially Hammarling's method. In our experience in Martin Köhler's PhD the Bartels-Steward method seemed to be more accurate and it has potential for shared memory parallelizatio. There is a drop in replacement we provided in a SLICOT Working Note and Martin is on an even faster OpenMP4 based implementation which is supposed to be published in his MEPACK at some point. So my suggestion would be to stick with the AD routine for now and check out MEPACK once it becomes available.

**Reply by artpelling** [reply]
I made a quick runtime comparison for `sb03od`. Unfortunately, I did not manage to get `sg03bd` to work, due to some weird LAPACK DGGES error that seems to be related to ldwork. 

`sb03od` looks faster, although I did not compare the precision. I hope I did not make any mistakes in the benchmark :^)

![pymor_vs_sb03od](https://user-images.githubusercontent.com/23401237/148282520-79adc91c-73a3-46cf-bbe4-7b450d4c38ba.png)

This is the code I used:
```
import numpy as np
import perfplot
from pymor.algorithms.lyapunov import solve_lyap_lrcf
from pymor.algorithms.to_matrix import to_matrix
from pymor.operators.numpy import NumpyMatrixOperator
from pymor.vectorarrays.numpy import NumpyVectorArray
import slycot


def wrap_sb03od(A, E, B):
    A = to_matrix(A)
    B = B.to_numpy()
    m, n = B.shape
    Q = np.zeros_like(A)
    U = np.zeros_like(A)
    U[:n, :m] = B.T
    out, _, _ = slycot.sb03od(n, m, A, Q, U, "C", trans="T")
    return out


# setup
n = 999

rng = np.random.RandomState(0)
A = np.eye(n)
B = rng.standard_normal((n, 5))

# benchmark E=None
out1 = perfplot.bench(
    setup=lambda n: (
        NumpyMatrixOperator(A[:n, :n]),
        None,
        NumpyVectorArray(B[:n].T, NumpyMatrixOperator(A[:n, :n]).source),
    ),
    kernels=[
        lambda x: solve_lyap_lrcf(*x, default_dense_solver_backend="slycot"),
        lambda x: wrap_sb03od(*x),
    ],
    labels=["Pymor", "Slycot"],
    n_range=np.round(np.geomspace(5, n, 20)).astype(int),
    xlabel="Problem size",
    equality_check=None,
)

# plotting
plot_opts = dict(logx=True)
out1.show(**plot_opts)

```

**Reply by lbalicki** [reply]
> @lbalicki if one would want to go down the one function for all lane, I would even suggest to use`sylv` rather than `lyap` in the name and prepare it for the most general Sylvester-type equation `A X B + C X D = E` of which both equations are special cases and other cases have applications in transfer function evaluation, rational Krylov subspaces, eigenvalue problems etc.

That's true and in my opinion it would just be messy since the docstring would be a very long and the `solve_sylv` function would need several extra arguments to distinguish between the equations of interest.
> 
> Regarding the SG03AD vs BD discussion: The first implements Bartels-Steward while the second is essentially Hammarling's method. In our experience in Martin Köhler's PhD the Bartels-Steward method seemed to be more accurate and it has potential for shared memory parallelizatio. There is a drop in replacement we provided in a SLICOT Working Note and Martin is on an even faster OpenMP4 based implementation which is supposed to be published in his MEPACK at some point. So my suggestion would be to stick with the AD routine for now and check out MEPACK once it becomes available.

With the extra error estimate that is comptued the sb03md/sg03ad are about an order of magnitude slower on my machine, similar to @artpelling's result. Since we only solve small systems with these methods it's not a big deal (in the vast majority of cases) in my opinion and the error estimate might be useful sometimes. Ideally, we have both methods available in pyMOR at some point.

**Reply by drittelhacker** [reply]
regarding `sole_sylv` that is very true and would probably lead to a lot of branching in the code as well to get the numerics to perform properly.

For the performance: I would still recommend revisiting this with MEPack once it is available. The tiled implementation can start paying of for moderate sizes already, leading to notable speedups due to multithreading and Martin uses tailored solvers for very small equations where he can eliminate several of the argument checks in the Slicot routines that start making up a significant fraction of the runtime, when the dimensions get small.

**Reply by artpelling** [reply]
> regarding `sole_sylv` that is very true and would probably lead to a lot of branching in the code as well to get the numerics to perform properly.
> 
> For the performance: I would still recommend revisiting this with MEPack once it is available. The tiled implementation can start paying of for moderate sizes already, leading to notable speedups due to multithreading and Martin uses tailored solvers for very small equations where he can eliminate several of the argument checks in the Slicot routines that start making up a significant fraction of the runtime, when the dimensions get small.

This sounds really interesting! Do you know if Python wrappers will be available and is there a planned release date?

In order to wrap up this discussion: I gather it would be best to create dedicated functions for the discrete equations and leave the Cholesky factorization as is. Should we make an issue out of this as a reminder for the future?

**Reply by drittelhacker** [reply]
> 
> This sounds really interesting! Do you know if Python wrappers will be available and is there a planned release date?
 I will discuss the details with Martin as soon as possible. The PhD was defended last year, codes should be ready, so, I guess, it is merely a matter of some polishing, the decision for a license and finding the time to actually do it..

> In order to wrap up this discussion: I gather it would be best to create dedicated functions for the discrete equations and leave the
> Cholesky factorization as is. Should we make an issue out of this as a reminder for the future?
A reminder can't hurt, I guess.
